tools/tcpbind.py

```diff
@@ -0,0 +1,463 @@
+                   skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+               flow_key.sport = sport;
+               ipv6_count.increment(flow_key);""",
+        'trace': """
+               struct ipv6_bind_data_t data6 = {.pid = pid, .ip = ipver};
+               data6.uid = bpf_get_current_uid_gid();
+               data6.ts_us = bpf_ktime_get_ns() / 1000;
+               bpf_probe_read(&data6.saddr, sizeof(data6.saddr),
+                   skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
+               data6.return_code = ret;
+               data6.sport = sport;
+               data6.bound_dev_if = skp->__sk_common.skc_bound_dev_if;
+               data6.socket_options = opts.data;
+               bpf_get_current_comm(&data6.task, sizeof(data6.task));
+               ipv6_bind_events.perf_submit(ctx, &data6, sizeof(data6));"""
+    },
+    'filter_cgroup': """
+    u64 cgroupid = bpf_get_current_cgroup_id();
+    if (cgroupset.lookup(&cgroupid) == NULL) {
+      return 0;
+    }""",
+}
+
+# code substitutions
+if args.count:
+    bpf_text = bpf_text.replace("IPV4_CODE", struct_init['ipv4']['count'])
+    bpf_text = bpf_text.replace("IPV6_CODE", struct_init['ipv6']['count'])
+else:
+    bpf_text = bpf_text.replace("IPV4_CODE", struct_init['ipv4']['trace'])
+    bpf_text = bpf_text.replace("IPV6_CODE", struct_init['ipv6']['trace'])
+
+if args.pid:
+    bpf_text = bpf_text.replace('FILTER_PID',
+        'if (pid != %s) { return 0; }' % args.pid)
+if args.port:
+    sports = [int(sport) for sport in args.port.split(',')]
+    sports_if = ' && '.join(['sport != %d' % sport for sport in sports])
+    bpf_text = bpf_text.replace('FILTER_PORT',
+        'if (%s) { currsock.delete(&pid); return 0; }' % sports_if)
+if args.uid:
+    bpf_text = bpf_text.replace('FILTER_UID',
+        'if (uid != %s) { return 0; }' % args.uid)
+if args.errors:
+    bpf_text = bpf_text.replace('FILTER_ERRORS', 'ignore_errors = 0;')
+if args.cgroupmap:
+    bpf_text = bpf_text.replace('FILTER_CGROUP', struct_init['filter_cgroup'])
+    bpf_text = bpf_text.replace(
+        'CGROUP_MAP',
+        (
+            'BPF_TABLE_PINNED("hash", u64, u64, cgroupset, 1024, "%s");' %
+            args.cgroupmap
+        )
+    )
+
+bpf_text = bpf_text.replace('FILTER_PID', '')
+bpf_text = bpf_text.replace('FILTER_PORT', '')
+bpf_text = bpf_text.replace('FILTER_UID', '')
+bpf_text = bpf_text.replace('FILTER_ERRORS', '')
+bpf_text = bpf_text.replace('FILTER_CGROUP', '')
+bpf_text = bpf_text.replace('CGROUP_MAP', '')
+
+# selecting output format - 80 characters or wide, fitting IPv6 addresses
+header_fmt = "%5s %-12.12s %-2s %-15s %-5s %5s %2s"
+output_fmt = b"%5d %-12.12s %-2d %-15.15s %5d %-5s %2d"
+if args.wide:
+    header_fmt = "%10s %-12.12s %-2s %-39s %-5s %5s %2s"
+    output_fmt = b"%10d %-12.12s %-2d %-39s %5d %-5s %2d"
+
+if args.ebpf:
+    print(bpf_text)
+    exit()
+
+if args.debug_source:
+    b = BPF(text=bpf_text, debug=DEBUG_SOURCE)
```

Comment:
Typically we do not have debug_source in the tools as this goes to assembly level and typically for tool writers. But I am okay with it if you feel you may use it when to modify the tools.

Reply:
I have removed debug-source printing.